### PR TITLE
[thread-cert] refactor case 5.3.2 using pktverify

### PR DESCRIPTION
**Pull request by FiveDimensions**

### CODE REVIEW

File: tests/scripts/thread-cert/Cert_5_3_02_RealmLocal.py [hunk 1]
```diff
@@ -84,46 +109,195 @@ def test(self):
...
+        for i in (1, 2):
+            pv.verify_attached('ROUTER_%d' % i)
+        pkts.filter_wpan_src64(SED).\
+            filter_wpan_dst64(ROUTER_2).\
+            filter_mle_cmd(MLE_CHILD_ID_REQUEST).\
+            must_next()
+        pkts.filter_wpan_src64(ROUTER_2).\
+            filter_wpan_dst64(SED).\
+            filter_mle_cmd(MLE_CHILD_ID_RESPONSE).\
+            must_next()
```
Comment (librasungirl) [inline tests/scripts/thread-cert/Cert_5_3_02_RealmLocal.py:164]: update verify_attached() to support both routers and end devices?

File: tests/scripts/thread-cert/Cert_5_3_02_RealmLocal.py [hunk 2]
```diff
@@ -84,46 +109,195 @@ def test(self):
...
+        _pkt = pkts.filter_ping_request().\
+            filter_ipv6_src_dst(LEADER_MLEID, ROUTER_2_MLEID).\
+            filter(lambda p: p.icmpv6.data.len == 256).\
+            must_next()
+        pkts.filter_ping_reply(identifier=_pkt.icmpv6.echo.identifier).\
+            filter_ipv6_src_dst(ROUTER_2_MLEID, LEADER_MLEID).\
+            filter(lambda p: p.icmpv6.data.len == 256).\
+            must_next()
+
+        # Step 3: Leader sends a Unfragmented ICMPv6 Echo Request to DUT’s
```
Comment (librasungirl) [inline tests/scripts/thread-cert/Cert_5_3_02_RealmLocal.py:178]:
```suggestion
        # Step 3: Leader sends a Unfragmented ICMPv6 Echo Request to DUT’s MLEID.
```

File: tests/scripts/thread-cert/Cert_5_3_02_RealmLocal.py [hunk 3]
```diff
@@ -84,46 +109,195 @@ def test(self):
...
+
+        # Step 6: Leader sends a Fragmented ICMPv6 Echo Request to the
+        #         Realm-Local All Routers multicast address (FF03::2)
+        #         The DUT MUST respond with an ICMPv6 Echo Reply
+        #         The DUT MUST NOT forward the ICMPv6 Echo Request to SED
+
+        _pkt1 = pkts.filter_ping_request().\
+            filter_wpan_src64(LEADER).\
+            filter_ipv6_dst(REALM_LOCAL_ALL_ROUTERS_ADDRESS).\
+            filter(lambda p: p.icmpv6.data.len == 256).\
```
Comment (librasungirl) [inline tests/scripts/thread-cert/Cert_5_3_02_RealmLocal.py:236]: avoid magic number `256` in this PR

File: tests/scripts/thread-cert/Cert_5_3_02_RealmLocal.py [hunk 4]
```diff
@@ -38,6 +38,7 @@
 ROUTER1 = 2
 DUT_ROUTER2 = 3
 SED1 = 4
+DATA_LEN = 256
```
Comment (librasungirl) [inline tests/scripts/thread-cert/Cert_5_3_02_RealmLocal.py:41]: nit: update the name to be more self-descriptive

File: tests/scripts/thread-cert/pktverify/packet_verifier.py [hunk 5]
```diff
@@ -161,29 +161,42 @@ def _add_initial_vars(self):
...
+            assert self.is_thread_device(parent), parent
+            src_pkts = pkts.filter_wpan_src64(child_extaddr).\
+                filter_wpan_dst64(self.vars[parent])
         src_pkts.filter_mle_cmd(MLE_CHILD_ID_REQUEST).must_next()  # Child Id Request
         result.record_last('child_id_request', pkts)
 
-        dst_pkts = pkts.filter_wpan_dst64(extaddr)
+        dst_pkts = pkts.filter_wpan_dst64(child_extaddr)
+        if parent:
+            src_pkts = pkts.filter_wpan_src64(self.vars[parent]).\
```
Comment (librasungirl) [inline tests/scripts/thread-cert/pktverify/packet_verifier.py:187]: typo:`src_pkts`->`dst_pkts`?

File: tests/scripts/thread-cert/pktverify/packet_verifier.py [hunk 6]
```diff
@@ -161,29 +161,42 @@ def _add_initial_vars(self):
...
+                filter_wpan_dst64(child_extaddr)
         dst_pkts.filter_mle_cmd(MLE_CHILD_ID_RESPONSE).must_next()  # Child Id Response
         result.record_last('child_id_response', pkts)
 
         with pkts.save_index():
-            src_pkts.filter_mle_cmd(MLE_ADVERTISEMENT).must_next()  # MLE Advertisement
-            result.record_last('mle_advertisement', pkts)
-            logging.info(f"verify attached: d={name}, result={result}")
+            if child_type:
+                assert child_type in ('FTD', 'MTD'), role
```
Comment (librasungirl) [inline tests/scripts/thread-cert/pktverify/packet_verifier.py:194]: what the value of `role` would be?

File: tests/scripts/thread-cert/pktverify/packet_verifier.py [hunk 7]
```diff
@@ -161,29 +161,42 @@ def _add_initial_vars(self):
...
+            if child_type != 'MTD':
+                src_pkts = pkts.filter_wpan_src64(child_extaddr)
+                src_pkts.filter_mle_cmd(MLE_ADVERTISEMENT).must_next()  # MLE Advertisement
+                result.record_last('mle_advertisement', pkts)
```
Comment (librasungirl) [inline tests/scripts/thread-cert/pktverify/packet_verifier.py:198]: Either child_type has a default value or indent this block under above `if child_type` range?

**Review by librasungirl — APPROVED** [review]

**Review by jwhui — APPROVED** [review]

**Pull request merged**